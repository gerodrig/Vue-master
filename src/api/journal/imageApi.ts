import { v2 as cloudinary } from 'cloudinary';

cloudinary.config({
    cloud_name: import.meta.env.VITE_CLOUDINARY_CLOUD_NAME,
    api_key: import.meta.env.VITE_CLOUDINARY_API_KEY,
    api_secret: import.meta.env.VITE_CLOUDINARY_API_SECRET,
    secure: true,
  });

export const deleteImage = async (url: string | null): Promise<boolean> => {

    if(!url) return false;

    if (url) {
        //url format https://res.cloudinary.com/gerodrig18/image/upload/v1682391219/vue-test/wbk4pifuthuyxqn08ly1.jpg

        // get vue-test/wbk4pifuthuyxqn08ly1
        // const public_id = url.split('/').pop()?.split('.')[0] || '';
        const public_id = url.split('/').slice(-2).join('/').split('.')[0] || '';
  
        try {
          const result = await cloudinary.api.delete_resources([public_id], {});
            //check if deleted: {'deleted': {'vue-test/': 'deleted'}} includes deleted return true
            //else return false
            if(Object.values(result.deleted)[0] === 'deleted') return true;
            
        } catch (error) {
          console.log(error);
        }
      }

      return false;
};