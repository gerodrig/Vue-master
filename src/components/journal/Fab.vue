<template>
<button
    @click="$emit('on:click')"
    class="fixed w-20 h-20 px-4 py-2 font-bold text-white bg-green-500 rounded-full bottom-5 right-5 hover:bg-green-600">
    <i :class="icon" class = "fas fa-2x"></i>
  </button>
</template>

<script lang='ts'>
export default {
  name: 'ComponentName',
  props: {
    icon: {
      type: String,
      default: 'fa-plus'
    }
  },
};
</script>