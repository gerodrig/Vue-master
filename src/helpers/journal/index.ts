export {default as uploadImage} from './uploadImage';
export { getDate } from "./getDate";


