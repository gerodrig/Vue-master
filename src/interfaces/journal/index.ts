export interface JournalState {
        isLoading: boolean;
        entries: Entry[];
}

export interface AuthState {
    status: AuthStatus;
    user: null | User;
    idToken: null | string;
    refreshToken: null | string;
}

export interface User {
    username?: string;
    email: string;
    password?: string;
}


//enum for the auth status
export type AuthStatus = 'authenticated' | 'not-authenticated' | 'authenticating';

export interface Rootstate {
    journal?: JournalState;
    auth?: AuthState;
}

export interface Entry {
    id: string;
    date: string | number;
    text: string;
    picture?: string | null;
}

export interface FormattedDate {
    day: number;
    month: string;
    year: number;
    dayOfWeek: string;
  }

export interface EntryFirebase {
    [ key: string ]: {
        date: string;
        text: string;
        picture: string | null;
    }
}