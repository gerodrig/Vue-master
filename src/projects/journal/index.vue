<template>
  <!-- <AuthPage /> -->
  <router-view />
  <!-- <div class="flex flex-col items-center justify-center h-screen text-3xl">
    <Logo />
    <h1 class="mb-4 text-5xl">Journal App</h1>
    <button
      class="px-4 py-2 mt-4 font-bold text-white bg-green-700 rounded hover:bg-green-800"
    >
      <RouterLink :to="{ name: 'no-entry' }">Click to launch App</RouterLink>
    </button>
  </div> -->
</template>

<script lang="ts">
import useAuth from '@/composables/auth/useAuth';
import { defineAsyncComponent, watch } from 'vue';
import { useRouter } from 'vue-router';

export default {
  name: 'Journal App',
  components: {
    Logo: defineAsyncComponent(() => import('@/components/shared/Logo.vue')),
    AuthPage: defineAsyncComponent(
      () => import('@/auth/journal/layouts/AuthLayout.vue')
    ),
  },
  setup() {
    const { checkAuthStatus, authStatus } = useAuth();
    const router = useRouter();

    watch(authStatus, () => {
      if(authStatus.value !== 'authenticated') {
        router.push({ name: 'login' });
      }
    });

    checkAuthStatus();
    return {
      authStatus,
    };
  },
};
</script>
