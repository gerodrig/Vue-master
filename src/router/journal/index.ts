import { RouteLocationNormalized as Route } from 'vue-router';

export const RouteNames = {
    HOME: 'daybook',
    ENTRY: 'entry',
    NOENTRY: 'no-entry',
  }

export default {
    name: RouteNames.HOME,
    component: () => import(/* webpackChunkName: "daybook" */ '@layouts/journal/DayBookLayout.vue'), 
    children: [
        {
            path: '',
            name: RouteNames.NOENTRY,
            component: () => import(/* webpackChunkName: "daybook-no-entry" */ '@/views/journal/NoEntrySelected.vue') 

        },
        {
            path: ':id',
            name: RouteNames.ENTRY,
            component: () => import(/* webpackChunkName: "daybook-entry" */ '@/views/journal/EntryView.vue'),
            props: (route: Route) => {
                return {
                    id: route.params.id
                }
            }
        }
    ],
}

