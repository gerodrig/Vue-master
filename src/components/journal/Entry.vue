<template>
  <div
    class="p-2 mb-3 transition-all duration-200 ease-in border-b border-green-600 cursor-pointer hover:bg-slate-300"
    @click="$router.push({ name: 'entry', params: { id } })"
  >
    <!-- Title -->
    <div class="flex">
      <span class="text-lg font-bold text-green-600">{{ day }}</span>
      <span class="mx-1 font-bold">{{ month }}</span>
      <span class="mx-2 font-light">{{ year }}, {{ dayOfWeek }}</span>
    </div>

    <div class="text-base">
      {{ text }}
    </div>
  </div>
</template>

<script lang="ts">
import { computed, toRef } from 'vue';
import { Entry, FormattedDate} from '@interfaces/journal/index';
import { getDate } from '../../helpers/journal/getDate';



export default {
  name: 'Entry',
  props: {
    entry: {
      type: Object as () => Entry,
      required: true,
    },
  },
  setup(props ) {
    const entry = toRef(props, 'entry');
    const date = computed<FormattedDate>(() => getDate(entry.value));

    const shortText = computed(() => {
      const text = entry.value.text;
      return text.length > 100 ? text.substring(0, 100) + '...' : text;
    });

    return {
      ...date.value,
      text: shortText,
      id: computed(() => entry.value.id),
    };
  },
};
</script>
