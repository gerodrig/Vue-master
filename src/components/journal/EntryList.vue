
<template>
    <div class="h-screen mx-4 mt-4 border-r border-green-900 entry-list-container">
        <input 
            type="text"
            class="items-center w-11/12 px-2 py-2 mx-auto rounded-md shadow-md ring-1 ring-gray-400 focus:border-green-500 focus:ring focus:ring-green-200 focus:ring-opacity-50"
            placeholder="Search Entry"
            v-model="term"
        />

        <div class="flex flex-col items-center justify-center my-4">
            <button @click="$router.push({name: 'entry', params: {id: 'new'}})" class="w-2/3 px-4 py-2 font-semibold text-white bg-green-800 rounded hover:bg-green-900 ">
                <i class="fas fa-plus-circle"></i>
                New Entry
            </button>
        </div>

        <div class="overflow-scroll entry-scroll-area">
            <Entry :entry="entry" v-for="entry in entries" :key="entry.id" />
        </div>
    </div>
</template>

<script lang="ts">
import { defineAsyncComponent, computed, ref, watch } from 'vue';
import { useStore } from 'vuex';
import { Rootstate, Entry } from '@interfaces/journal/index';

export default {
    name: 'EntryList',
    components: {
        Entry: defineAsyncComponent(() => import('@components/journal/Entry.vue'))
    },
    setup() {
        const store = useStore<Rootstate>();
        const term = ref('');
        const getEntriesByTerm = (value: string): Entry[] =>  store.getters['journal/getEntriesByTerm'](value);
        
        const entries =  computed<Entry[]>(() => {
            if (term.value.length > 2) {
                return getEntriesByTerm(term.value);
            } else {
                return store.state.journal.entries;
            }
        });
        
        return {
            term,
            entries
        }
    }
}
</script>

<style   scoped >
.entry-list-container {
    height: calc(100vh - 56px);
}

.entry-scroll-area {
    height: calc( 100vh - 120px);
}

</style>