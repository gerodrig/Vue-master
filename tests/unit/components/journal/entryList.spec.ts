import { describe, it, vi, expect, beforeEach } from 'vitest';
import { VueWrapper, shallowMount } from '@vue/test-utils';

import { createStore } from 'vuex';

import EntryList from '@/components/journal/EntryList.vue';
import journal from '../../mocks/journal/test-journal-state';
import { JournalState } from '../../../../src/interfaces/journal/index';

//?USE THE REAL STORE
const createVuexStore = ( initialState: JournalState) => createStore({
    modules: {
        journal: {
            ...journal,
            state: { ...initialState }
        }
    }
});

describe('EntryList component tests', () => {

    //configure the store with the journal module
    // const journalMockModule = {
    //     namespaced: true,
    //     getters: {
    //         getEntriesByTerm
    //     },
    //     state: () => ({
    //         isLoading: false,
    //         entries: journal.entries
    //     })
    // };

    // const store = createStore({
    //     modules: {
    //         journal: {...journalMockModule}
    //     }
    // });


    //?USE THE REAL STORE
    const store = createVuexStore(journal);

    //router mock
    const mockRouter = {
        push: vi.fn()
    }

    let wrapper: VueWrapper<any>;
    
    //restart wrapper before each test
    beforeEach(() => {
        vi.clearAllMocks();
        wrapper = shallowMount(EntryList, {
            global: {
                mocks: {
                $router: mockRouter
                },
            plugins: [store],
            },
        });
    
    });

  it('should call getEntriesByTerm and show 5 entries.', () => {

    expect(wrapper.findAll('entry-stub').length).toBe(journal.entries.length);
    expect(wrapper.html()).toMatchSnapshot();

  });

  it('should call getEntriesByTerm and filter entries.', async () => {
       
    const input = wrapper.find('input');
    // await input.setValue('chi');

    // expect(wrapper.findAll('entry-stub').length).toBe(2);
    expect( input.exists() ).toBeTruthy();

  })

  it('buttons new shoudl redirect to /new', () => {
    const button = wrapper.find('button');
    button.trigger('click');

    expect(mockRouter.push).toHaveBeenCalled();
    expect(mockRouter.push).toHaveBeenCalledWith({name: 'entry', params: {id: 'new'}});
  });
});