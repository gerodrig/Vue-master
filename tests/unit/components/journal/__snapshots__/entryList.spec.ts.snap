// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`EntryList component tests > should call getEntriesByTerm and show 5 entries. 1`] = `
"<div data-v-cac28731=\\"\\" class=\\"h-screen mx-4 mt-4 border-r border-green-900 entry-list-container\\"><input data-v-cac28731=\\"\\" type=\\"text\\" class=\\"items-center w-11/12 px-2 py-2 mx-auto rounded-md shadow-md ring-1 ring-gray-400 focus:border-green-500 focus:ring focus:ring-green-200 focus:ring-opacity-50\\" placeholder=\\"Search Entry\\">
  <div data-v-cac28731=\\"\\" class=\\"flex flex-col items-center justify-center my-4\\"><button data-v-cac28731=\\"\\" class=\\"w-2/3 px-4 py-2 font-semibold text-white bg-green-800 rounded hover:bg-green-900\\"><i data-v-cac28731=\\"\\" class=\\"fas fa-plus-circle\\"></i> New Entry </button></div>
  <div data-v-cac28731=\\"\\" class=\\"overflow-scroll entry-scroll-area\\">
    <entry-stub data-v-cac28731=\\"\\" entry=\\"[object Object]\\"></entry-stub>
    <entry-stub data-v-cac28731=\\"\\" entry=\\"[object Object]\\"></entry-stub>
    <entry-stub data-v-cac28731=\\"\\" entry=\\"[object Object]\\"></entry-stub>
    <entry-stub data-v-cac28731=\\"\\" entry=\\"[object Object]\\"></entry-stub>
    <entry-stub data-v-cac28731=\\"\\" entry=\\"[object Object]\\"></entry-stub>
  </div>
</div>"
`;
